### PR TITLE
Bug 1253656 - Enable bookmark keyword searching

### PR DESCRIPTION
**Pull request by mauryat**
@thebnich let me know if you have any feedback! 

My main concern is:
What happens if the user submits a keyword search query. Even before the database request gets completed, it is possible that the user might edit the url bar text and resubmit.

### CODE REVIEW

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 1]
```diff
@@ -1557,17 +1557,49 @@ extension BrowserViewController: URLBarDelegate {
...
 
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        profile.bookmarks.modelFactory >>== {
+            let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ")!
+            let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+            let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+            
+            $0.getURLForKeywordSearch(possibleKeyword)
+            .uponQueue(dispatch_get_main_queue()) {
+                if let urlString = $0.successValue {
```
Comment (mauryat) [inline Client/Frontend/Browser/BrowserViewController.swift]: I've noticed that this line sometimes gets hit twice (even though I haven't re-edited and re-submitted the text). Haven't been able to reproduce that again.

Comment (mauryat) [inline Client/Frontend/Browser/BrowserViewController.swift]: I've noticed that this line sometimes gets hit twice (even though I haven't re-edited and re-submitted the text). Haven't been able to reproduce that again.

File: Storage/Bookmarks/BookmarksModel.swift [hunk 2]
```diff
@@ -134,6 +134,7 @@ public protocol BookmarksModelFactory {
     var nullModel: BookmarksModel { get }
 
     func isBookmarked(url: String) -> Deferred<Maybe<Bool>>
+    func getURLForKeywordSearch(keyword: String) -> Deferred<Maybe<String>>
```
Comment (rnewman) [inline Storage/Bookmarks/BookmarksModel.swift]: I think this should not be on `BookmarksModel`. Instead, put it on a new protocol — perhaps `KeywordSearchSource` — to which `SQLiteBookmarks` conforms and which `Profile.bookmarks` requires.

Comment (mauryat) [inline Storage/Bookmarks/BookmarksModel.swift]: @rnewman I fixed this. Though, it would help to understand why this is better. Is it because we don't care about the concept of _keyword searching_ in the `BookmarksModel` context?

File: Storage/SQL/SQLiteBookmarksModel.swift [hunk 3]
```diff
@@ -855,4 +890,10 @@ public class MergedSQLiteBookmarks: BookmarksModelFactorySource {
         self.local = SQLiteBookmarks(db: db)
         self.buffer = SQLiteBookmarkBufferStorage(db: db)
     }
+
+    public func getURLForKeywordSearch(keyword: String) -> Deferred<Maybe<String>> {
+        return modelFactory >>== {
+            return ($0 as! KeywordSearchSource).getURLForKeywordSearch(keyword)
```
Comment (mauryat) [inline Storage/SQL/SQLiteBookmarksModel.swift]: @rnewman I wish I could avoid downcasting and instead, declare something like:
public var modelFactory: protocol<Deferred<Maybe<BookmarksModelFactory>>, KeywordSearchSource>

Is that possible?

Comment (rnewman) [inline Storage/SQL/SQLiteBookmarksModel.swift]: I suppose I should rephrase this as "keyword searching is nothing to do with Model".

`BookmarksModel` exists for one purpose: to back a hierarchical view of bookmark folders.

Other things about bookmarks live in other protocols.

File: Storage/Bookmarks/BookmarksModel.swift (hunk 2 above)
Comment (rnewman) [inline Storage/Bookmarks/BookmarksModel.swift]: Yes, exactly.

File: Storage/SQL/SQLiteBookmarksModel.swift (hunk 3 above)
Comment (rnewman) [inline Storage/SQL/SQLiteBookmarksModel.swift]: If you make `SQLiteBookmarks` implement `KeywordSearchSource`, and put the logic there rather than in `SQLiteBookmarksModelFactory`, then this simply becomes `return self.local.getURLForKeywordSearch(keyword)`.

Comment (mauryat) [inline Storage/SQL/SQLiteBookmarksModel.swift]: @rnewman fixed this. Btw, what are the mirror tables for?

Comment (rnewman) [inline Storage/SQL/SQLiteBookmarksModel.swift]: If you really want to know:

https://gist.github.com/rnewman/82c17460161047db38d5

Comment (mauryat) [inline Storage/SQL/SQLiteBookmarksModel.swift]: Great thank you! Any other feedback on the code? Or, is it good to go?

File: Storage/Bookmarks/BookmarksModel.swift [hunk 4]
```diff
@@ -332,7 +332,7 @@ public class MockMemoryBookmarksStore: BookmarksModelFactory, ShareToDestination
     public func isBookmarked(url: String) -> Deferred<Maybe<Bool>> {
         return deferMaybe(DatabaseError(description: "Not implemented"))
     }
-
+
```
Comment (rnewman) [inline Storage/Bookmarks/BookmarksModel.swift]: You can remove this change.

File: Storage/SQL/SQLiteBookmarksBase.swift [hunk 5]
```diff
@@ -35,4 +45,21 @@ public class SQLiteBookmarks: BookmarksModelFactorySource {
         
         return self.db.queryReturnsResults(sql, args: args)
     }
+
+    public func getURLForKeywordSearch(keyword: String) -> Deferred<Maybe<String>> {
+        let sql = "SELECT bmkUri FROM \(TableBookmarksBuffer) WHERE " +
```
Comment (rnewman) [inline Storage/SQL/SQLiteBookmarksBase.swift]: I don't think querying the buffer is the correct choice here.

The buffer will — as soon as bidirectional syncing is enabled — be empty. That this code works now is almost an accident!

Because we don't allow you to edit bookmarks locally, or add keywords to them locally, an acceptable short-term solution here is to query the overlay of remote bookmarks on top of merged bookmarks (`ViewBoomarksBufferOnMirror`). A longer-term solution is to query everything: use `ViewAllBookmarks`.

Those views don't pass through `keyword` (because we haven't used it so far), so you will need a small database migration to do what you need. Ask @sleroux for help with that.

**Comment by mauryat** [comment]
@thebnich, is there a test firefox sync account I can use in the XCUI test case that I'm writing?

File: Storage/SQL/SQLiteBookmarksBase.swift (hunk 5 above)
Comment (mauryat) [inline Storage/SQL/SQLiteBookmarksBase.swift]: @sleroux, can you help me with the database migration? I think we need to add an ALTER TABLE command to add the new columns to the migration script.

Comment (sleroux) [inline Storage/SQL/SQLiteBookmarksBase.swift]: Sure thing. Since you're going to be modifying a SQL VIEW instead of a SQL TABLE you won't need to do an ALTER TABLE.

As Richard mentioned, we already have the `keyword` field in the bookmark tables [1] but we don't have them part of the `ViewBoomarksBufferOnMirror` SQL VIEW [2]. Since VIEWs are read-only and are queries on top of existing tables, instead of altering it, you'll need to just `DROP` the VIEW and recreate it. We've done this in a few places already that you can use as reference [3]. As far as where the migration code should live, you'll want to add another `if` case to the `updateTable` in `BrowserTable` and update the `DefaultVersion` to DefaultVersion+1 along with an associated bug number [4]

[1] https://github.com/mozilla-mobile/firefox-ios/blob/master/Storage/SQL/BrowserTable.swift#L299
[2] https://github.com/mozilla-mobile/firefox-ios/blob/master/Storage/SQL/BrowserTable.swift#L380
[3] https://github.com/mozilla-mobile/firefox-ios/blob/master/Storage/SQL/BrowserTable.swift#L778
[4] https://github.com/mozilla-mobile/firefox-ios/blob/master/Storage/SQL/BrowserTable.swift#L110

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 6]
```diff
@@ -1540,17 +1540,46 @@ extension BrowserViewController: URLBarDelegate {
...
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ")!
+        let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+        let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+        
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) {
+                if let urlString = $0.successValue {
+                    let engine = self.profile.searchEngines.defaultEngine
 
+                    guard let url = engine.searchURLForQuery(possibleQuery, searchTemplate: urlString) else {
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: I'm confused--what's going on here? At this point, we've already called `getURLForKeywordSearch`, and we have a URL with the keyword engine (`urlString`). Why would we pass the URL from the keyword as a query to the default search engine?

File: Client/Frontend/Browser/OpenSearch.swift [hunk 7]
```diff
@@ -63,7 +64,8 @@ class OpenSearchEngine: NSObject, NSCoding {
     /**
      * Returns the search URL for the given query.
      */
-    func searchURLForQuery(query: String) -> NSURL? {
+    func searchURLForQuery(query: String, searchTemplate: String? = nil) -> NSURL? {
```
Comment (thebnich) [inline Client/Frontend/Browser/OpenSearch.swift]: I don't understand the changes in this file. If we're doing a keyword "search", we should get the URL template from the bookmarks, then replace the `%s` from the bookmark URL with the search term. We shouldn't need to make any changes to search engines.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 8]
```diff
@@ -1540,17 +1540,46 @@ extension BrowserViewController: URLBarDelegate {
...
             return
         }
 
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        let possibleKeywordQuerySeparatorSpace = trimmedText.characters.indexOf(" ")!
+        let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+        let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+        
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) {
+                if let urlString = $0.successValue {
```
Comment (mauryat) [inline Client/Frontend/Browser/BrowserViewController.swift]: @thebnich, I was told that Mozilla devs use a linting tool for swift. Which one is it? There seem to be quite a few such tools available. I tried [SwiftLint](https://github.com/realm/SwiftLint), but it seems to be running only a [small subset of the rules](https://github.com/realm/SwiftLint#rules). Also, these rules are based on [Github's style guide](https://github.com/github/swift-style-guide) and firefox uses a [different one](https://github.com/mozilla-mobile/firefox-ios#swift-style).

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 6 above)
Comment (mauryat) [inline Client/Frontend/Browser/BrowserViewController.swift]: @thebnich : Yes, at this point we need to replace `%s` in URL with the query. However, the problem is when the query contains whitespace. This should get replaced as `%20`.

I used the `OpenSearchEngine.searchURLForQuery()` function because it was already doing something similar (like percent encoding of unallowed characters). If there are characters (not just whitespace) that need to be percent encoded, either in the URL or in the query, they'll be handled here.

Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: I think the function you're talking about (`getURLFromTemplate`) does way more than we need -- it should just be a simple encoding of the search terms. Something like this should be enough:

`let escapedQuery = query.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.SearchTermsAllowedCharacterSet())`

(`SearchTermsAllowedCharacterSet` should be moved to `NSCharacterSetExtensions.swift`).

Keyword searching isn't related to search engines, so we should probably avoid using it here.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 9]
```diff
@@ -1541,17 +1541,45 @@ extension BrowserViewController: URLBarDelegate {
...
             return
         }
 
-        Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
+        let possibleKeyword = trimmedText.substringToIndex(possibleKeywordQuerySeparatorSpace)
+        let possibleQuery = trimmedText.substringFromIndex(possibleKeywordQuerySeparatorSpace.successor())
+        
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) { result in
+                if let urlString = result.successValue {
+                    let engine = self.profile.searchEngines.defaultEngine
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Same here -- arbitrarily using a different engine to morph it with a different template is a red flag that we're doing something wrong, I think. I suspect this would all be simpler if we just inlined the encoding.

Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: (My other comment is a bit buried; see it [here](https://github.com/mozilla-mobile/firefox-ios/pull/2123#discussion_r89912279).)

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 10]
```diff
@@ -1541,13 +1541,41 @@ extension BrowserViewController: URLBarDelegate {
...
+
+        profile.bookmarks.getURLForKeywordSearch(possibleKeyword).uponQueue(dispatch_get_main_queue()) { result in
+            if var urlString = result.successValue,
+                let escapedQuery = possibleQuery.stringByAddingPercentEncodingWithAllowedCharacters(NSCharacterSet.SearchTermsAllowedCharacterSet()),
+                let range = urlString.rangeOfString("%s") {
+                urlString.replaceRange(range, with: escapedQuery)
+
+                if let url = NSURL(string: urlString) {
+                    let engine = self.profile.searchEngines.defaultEngine
+                    Telemetry.recordEvent(SearchTelemetry.makeEvent(engine: engine, source: .URLBar))
```
Comment (thebnich) [inline Client/Frontend/Browser/BrowserViewController.swift]: Please remove these two lines -- this isn't a search on a search engine, so we shouldn't record it as one (we used to erroneously record URL entries as searches, but this was just fixed today in [bug 1319572](https://bugzilla.mozilla.org/show_bug.cgi?id=1319572)).

**Comment by thebnich** [comment]
Thanks for the PR and all the updates. This looks good! 👍

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 10 above)
Comment (mauryat) [inline Client/Frontend/Browser/BrowserViewController.swift]: @thebnich, removed those 2 lines. Also updated the order in which the parsing of the text happens. If it's a URL, I felt it would make sense to submit it immediately. Rather than wait for keyword lookup.

**Comment by mauryat** [comment]
@thebnich, looks like these haven't been "merged" yet. Just wanted to check if there was anything else to be done.

**Comment by mauryat** [comment]
@thebnich, I'm not able to merge the PR even after it's approved. I guess there's a restriction for contributors?

**Comment by thebnich** [comment]
Hrm OK, I thought we had fixed that. Thanks for the ping!
